dbt/include/trino/macros/adapters.sql

```diff
@@ -290,3 +290,20 @@
 
   {{ return({'relations': [target_relation]}) }}
 {% endmacro %}
+
+{% macro trino__alter_column_type(relation, column_name, new_column_type) %}
+  {#
+    1. Create a new column (w/ temp name and correct type)
+    2. Copy data over to it
+    3. Drop the existing column (cascade!)
+    4. Rename the new column to existing column
+  #}
+  {%- set tmp_column = column_name + "__dbt_alter" -%}
+
+  {% call statement('alter_column_type') %}
+    alter table {{ relation }} add column {{ adapter.quote(tmp_column) }} {{ new_column_type }};
+    update {{ relation }} set {{ adapter.quote(tmp_column) }} = CAST({{ adapter.quote(column_name) }} AS {{ new_column_type }});
+    alter table {{ relation }} drop column {{ adapter.quote(column_name) }};
+    alter table {{ relation }} rename column {{ adapter.quote(tmp_column) }} to {{ adapter.quote(column_name) }}
```

Comment:
This macro could be implemented as direct change of column datatype instead of creating temporary column:

```sql
alter table {{ relation }} alter column {{ adapter.quote(column_name) }} set data type {{ new_column_type }};
```
But this is somewhat limited in terms of the types that can be changed., e.g. 
```
Failed to set column type: Cannot change column type: some_int: long -> double
...long -> int
...long -> string
...long -> float
...long -> double
...int -> string
...double -> float
...
```

But also with current implementation we could have error e.g. when casting `bigint` to `int` and some value is out of range for integer, then we are left with `__dbt_alter` suffixed column.

But I'm ok with current implementation and just wanted to show this alternative approach. 
WDYT @posulliv @hovaesco @mdesmet

Reply:
I think the CAST is the most permissible approach. Let's go with that.

Reply:
yeah we've found the CAST to be more useful.

We considered the `set data type` option in `alter table` as well but we are using Iceberg tables and like you mentioned Iceberg limits what data type changes are possible. The only data type changes we found supported with Iceberg are:

```
FROM_TYPE  || TARGET_TYPE
=========================
integer    || integer OR long
float      || float OR double
decimal    || decimal (target must be same scale as from and precision of target must be greater than or equal to from's precision)
```